design/20230726-cel-policy.md

```diff
@@ -0,0 +1,418 @@
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of server certificates. Since the cluster is multi-tenant, I want to ensure that tenants do not
+"trip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
+ends with `<namespace>.<domain>`.
+
+A variant of this user story is exemplified in [#62](https://github.com/cert-manager/approver-policy/issues/62),
+and I think it's one of the most common user stories the proposed feature will solve.
+
+#### SPIFFE ID identifying namespace
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of [X.509 SVID certificates](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md).
+The [SPIFFE ID](https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE-ID.md) (X.509 URI SAN) should identify
+the namespace (and service account), and must start with `spiffe://<trust-domain>/ns/<namespace>/sa/`.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go into as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes/PKI ecosystem.
+
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposal is to add sibling properties to `values` and `value` properties inside the `allowed` block. The new
+fields should contain CEL validations used to validate CSR attributes with similar
+[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has
+today, but with the following extension: If validations are present for the given CSR attribute, and **ALL**
+validations passes, the attribute will be allowed. This also holds if no values are listed.
+
+The validations themselves should be expressed using the same API as Kubernetes `x-kubernetes-validations`, having
+a mandatory `rule` field for the CEL expression and an optional `message` field for the validation error message.
+
+To use validations in CertificateRequestPolicy the API may look like this (example):
+
+```yaml
+spec:
+  allowed:
+    dnsNames:
+      validations:
+        - message: only dnsNames in the local namespace service domain is allowed.
+          rule: self.endsWith(cr.namespace + '.svc') || self.endsWith(cr.namespace + '.svc.cluster.local')
```

Comment:
I would start with a `cr` value that only contains the basic fields of the CertifiicateRequest resource (name, namespace, annotations, labels, uuid, ....). Once that works, we could also add other properties (like x509 stuff).

Reply:
Agreed, I think we should start even simpler in the initial implementation and provide just CR `namespace` and `name`. I will add this initial MVP to the proposed design.

Reply:
Updated, PTAL!